tezos/new_context/src/working_tree/mod.rs

```diff
@@ -86,66 +57,169 @@ pub struct Commit {
+                    self.inner.set(inner);
+                };
                 Ok(hash_id)
             }
         }
     }
 
+    pub fn new(node_kind: NodeKind, entry: Entry) -> Self {
+        Node {
+            inner: Cell::new(
+                NodeInner::new()
+                    .with_commited(false)
+                    .with_node_kind(node_kind)
+                    .with_entry_hash_id(0)
+                    .with_entry_available(true)
+                    .with_entry_id(match entry {
+                        Entry::Tree(tree_id) => tree_id.into(),
+                        Entry::Blob(blob_id) => blob_id.into(),
+                        Entry::Commit(_) => panic!(), // A Node never contains a commit
+                    }),
+            ),
+        }
+    }
+
+    pub fn new_commited(
+        node_kind: NodeKind,
+        hash_id: Option<HashId>,
+        entry: Option<Entry>,
+    ) -> Self {
+        Node {
+            inner: Cell::new(
+                NodeInner::new()
+                    .with_commited(true)
+                    .with_node_kind(node_kind)
+                    .with_entry_hash_id(hash_id.map(|h| h.as_u32()).unwrap_or(0))
+                    .with_entry_available(entry.is_some())
+                    .with_entry_id(match entry {
+                        Some(Entry::Tree(tree_id)) => tree_id.into(),
+                        Some(Entry::Blob(blob_id)) => blob_id.into(),
+                        Some(Entry::Commit(_)) => panic!(), // A Node never contains a commit
+                        None => 0,
+                    }),
+            ),
+        }
+    }
+
     pub fn get_hash_id(&self) -> Result<HashId, MerkleError> {
-        self.entry_hash
-            .get()
+        self.hash_id()
             .ok_or(MerkleError::InvalidState("Missing entry hash"))
     }
 
     pub fn set_entry(&self, entry: &Entry) -> Result<(), MerkleError> {
-        self.entry
-            .try_borrow_mut()
-            .map_err(|_| MerkleError::InvalidState("The Entry is borrowed more than once"))?
-            .replace(entry.clone());
+        let mut inner = self.inner.get();
+
+        match entry {
+            Entry::Tree(tree_id) => {
+                let tree_id: u64 = (*tree_id).into();
+                if tree_id != inner.entry_id() || !inner.entry_available() {
+                    inner.set_entry_available(true);
+                    inner.set_node_kind(NodeKind::NonLeaf);
+                    inner.set_entry_id(tree_id);
+                }
+            }
+            Entry::Blob(blob_id) => {
+                let blob_id: u64 = (*blob_id).into();
+                if blob_id != inner.entry_id() || !inner.entry_available() {
+                    inner.set_entry_available(true);
+                    inner.set_node_kind(NodeKind::Leaf);
+                    inner.set_entry_id(blob_id);
+                }
+            }
+            Entry::Commit(_) => {
+                panic!()
```

Comment:
Same here. But are we 100% sure this can never be reached under any circumstance?

Reply:
As far as I know, a `Node` contains only a `Tree` or a `Blob`.
The enum `Entry` has a `Commit` member only to be able to (de)serialize easily between those 3 types.
I could remove `Commit` from `Entry` but then we would need 2 new functions: `serialize_commit` and `deserialize_commit`